### PR TITLE
Pre-work for the design change: create own python class for every model

### PR DESCRIPTION
**Pull request by vinnamkim**
### Summary

- This is the pre-work, so that we will actually introduce own python class for every model in the next PR.
- Enrich `OTXModel` to have `export()`, `register_explain_hook()`, and `label_info` property.
- `label_info` will be automatically constructed by the input integer argument, `num_classes`.

### Checklist

- [ ] I have added unit tests to cover my changes.​
- [ ] I have added integration tests to cover my changes.​
- [ ] I have added e2e tests for validation.
- [ ] I have added the description of my changes into CHANGELOG in my target branch (e.g., [CHANGELOG](https://github.com/openvinotoolkit/training_extensions/blob/develop/CHANGELOG.md) in develop).​
- [ ] I have updated the documentation in my target branch accordingly (e.g., [documentation](https://github.com/openvinotoolkit/training_extensions/tree/develop/docs) in develop).
- [ ] I have [linked related issues](https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword).

### License

- [ ] I submit _my code changes_ under the same [Apache License](https://github.com/openvinotoolkit/training_extensions/blob/develop/LICENSE) that covers the project.
      Feel free to contact the maintainers if that's a concern.
- [ ] I have updated the license header for each file (see an example below).

```python
# Copyright (C) 2023 Intel Corporation
# SPDX-License-Identifier: Apache-2.0
```

### CODE REVIEW

**Review by MarkByun — APPROVED** [review]
It looks good.

File: src/otx/core/data/dataset/base.py [hunk 1]
```diff
@@ -30,15 +30,27 @@
 
 
 @dataclass
-class DataMetaInfo:
-    """Meta information of each subset datasets."""
+class LabelInfo:
+    """Object to represent label information."""
```
Comment (jaegukhyun) [inline src/otx/core/data/dataset/base.py:34]: What if LitModule requires other info such as data_root, subset names? I'm not sure whether there are such use cases, but if there are, should we create DatasetMetaInfo class which have LabelInfo attribute?

File: src/otx/core/utils/config.py [hunk 2]
```diff
@@ -66,3 +67,29 @@ def mmconfig_dict_to_dict(obj: MMConfig | list[MMConfig]) -> list | dict:
         return {k: mmconfig_dict_to_dict(v) for k, v in obj.to_dict().items()}
 
     return obj
+
+
+def inplace_num_classes(
```
Comment (jaegukhyun) [inline src/otx/core/utils/config.py:72]: It seems this function is duplicated with https://github.com/openvinotoolkit/training_extensions/blob/0e8bd95b41e899c01ec2fee2a6258d24f6f08598/src/otx/core/utils/build.py#L82
Can we choose one of them?

File: src/otx/core/model/entity/base.py [hunk 3]
```diff
@@ -5,29 +5,66 @@
...
+
+        if old_num_classes != new_num_classes:
+            msg = (
+                f"Given LabelInfo has the different number of classes "
+                f"({old_num_classes}!={new_num_classes}). "
+                "The model prediction layer is reset to the new number of classes "
+                f"(={new_num_classes})."
+            )
+            warnings.warn(msg, stacklevel=0)
+            self._reset_prediction_layer(num_classes=label_info.num_classes)
```
Comment (jaegukhyun) [inline src/otx/core/model/entity/base.py:64]: It seems all of model entity do not implement their own _reset_prediciton_layer function. 
Do we have a plan for this?

**Review by sungmanc — APPROVED** [review]
Approved, BTW, could you please explain why do we need to insert the `num_classes` to the `OTXModel` at this PR ? I'm not opposite to inject this argument, however, just curiosity. It seems that inserting the `num_classes` have no much relation with own python class.

File: src/otx/core/utils/config.py (hunk 2 above)
Comment (vinnamkim) [reply src/otx/core/utils/config.py:72]: Please deprecate it after merging it.
https://github.com/openvinotoolkit/training_extensions/blob/0e8bd95b41e899c01ec2fee2a6258d24f6f08598/src/otx/core/utils/build.py#L89
This should be `isinstance(value, ListConfig)`.

File: src/otx/core/data/dataset/base.py (hunk 1 above)
Comment (vinnamkim) [reply src/otx/core/data/dataset/base.py:34]: We can rename and expand it if that case happens. Honestly, we already have the expanded one, `HLabelInfo`.

File: src/otx/core/model/entity/base.py (hunk 3 above)
Comment (vinnamkim) [reply src/otx/core/model/entity/base.py:64]: Sure we might do in the future

**Review by jaegukhyun — APPROVED** [review]

**Comment by vinnamkim** [comment]
> Approved, BTW, could you please explain why do we need to insert the `num_classes` to the `OTXModel` at this PR ? I'm not opposite to inject this argument, however, just curiosity. It seems that inserting the `num_classes` have no much relation with own python class.

We introduce `LabelInfo` as a common property of `OTXModel`. However, this property affects the shape of the model itself. If we decide it lazily, not at the construction, we should implement this `_reset_prediction_layer()` as well this time. In addition, lazily resetting the layer seems not good in terms of the clean flow and computation cost.

**Pull request merged**